src/Model3/NetDeamon.Model3/Internal/EntityStateCache.cs

```diff
@@ -1,23 +1,39 @@
-﻿using System.Collections.Concurrent;
+﻿using System;
+using System.Collections.Concurrent;
 using System.Diagnostics.CodeAnalysis;
+using System.Reactive.Linq;
+using System.Reactive.Subjects;
 using JoySoftware.HomeAssistant.Client;
 using JoySoftware.HomeAssistant.Model;
 
 namespace NetDaemon.Model3.Internal
 {
     [SuppressMessage("", "CA1812", Justification = "Is Loaded via DependencyInjection")]
-    internal class EntityStateCache
+    internal class EntityStateCache : IDisposable
     {
         private readonly IHassClient _hassClient;
+        private readonly ConcurrentDictionary<string, HassState?> _latestStates = new();
+        private readonly Subject<HassStateChangedEventData> _innerSubject = new();
 
-        public EntityStateCache(IHassClient hassClient)
+        public EntityStateCache(IHassClient hassClient, IObservable<HassEvent> events)
         {
             _hassClient = hassClient;
-        }
 
-        private readonly ConcurrentDictionary<string, HassState?> _latestStates = new();
+            var stateChanges = events
+                .Select(e => e.Data as HassStateChangedEventData)
+                .Where(e => e != null)
+                .Select(e => e!);
```

Comment:
Wondering if this is cleaner :


```suggestion
                .Where(e => e.Data is HassStateChangedEventData)
                .Cast<HassStateChangedEventData>();
```

I think that works because the `is` returns false on `null`.